FEATURE: Integrate conflict resolution with publish/discard dialog workflow

grebaldi opened the pull request:
https://github.com/neos/neos-ui/assets/2522299/5177f375-a586-4c97-8511-a9b60fd4b7a9

solves: https://github.com/neos/neos-ui/issues/3761
fixes: https://github.com/neos/neos-ui/issues/3845
fixes: #3785
builds on: https://github.com/neos/neos-ui/pull/3762

### The Problem

With the introduction of conflict resolution, users can synchronize their workspace with recent changes in the base workspace by clicking the "sync"-button that is placed aside the publish dropdown. When a workspace is in `OUTDATED` state, the "sync"-button shows up.

It may happen though that a workspace transitions into `OUTDATED` state without the UI taking notice of it. If the user then tries to publish their changes, publishing fails with an exception. Only after reloading the entire UI, they'll be able to trigger the "sync"-workflow.

**Steps to reproduce**

Assuming you have two user accounts with `Neos.Neos:Editor` role (let's call them "editor-a" and "editor-b"):

1. Log in as "editor-a"
2. Create a document "Test"
3. Add a text node to "Test"
4. Publish those changes
5. Edit the text node, without publishing that change
6. Log in as "editor-b" in a private window (keep "editor-a"'s window open!)
7. Sync "editor-b"'s workspace, so you can see document "Test"
8. Remove document "Test"
9. Publish that change
10. Switch to "editor-a"'s window (do not reload the page!)
11. Try to publish the change on the text node that is still pending
12. Observe:
  - **Pre-PR:** Publishing results in an exception
  - **Post-PR:** Publishing triggers conflict resolution

### The solution

With this PR, the UI automatically triggers the sync and conflict resolution workflows if a publishing operation fails due to an outdated workspace.

### Remaining TODOs

- [x] Move `Conflicts` concept to `Application\Shared` namespace
  - [x] Move `ConflictsOccurred` to `Application\Shared` namespace
  - [x] Move `Conflict` to `Application\Shared` namespace
  - [x] Move `Conflicts` to `Application\Shared` namespace
  - [x] Move `ConflictsBuilder` to `Application\Shared` namespace
  - [x] Move `ReasonForConflict` to `Application\Shared` namespace
  - [x] Move `TypeOfChange` to `Application\Shared` namespace
  - [x] Move `IconLabel` to `Application\Shared` namespace
- [x] Add command handlers for all Publish-related commands & make sure they throw `ConflictsOccurred`
  - [x] Move `PublishChangesInDocument` -> `PublishChangesInDocument\PublishChangesInDocumentCommand`
  - [x] Create `PublishChangesInDocumentCommandHandler`
  - [x] Ensure that `PublishChangesInDocumentCommandHandler` throws `ConflictsOccurred`
  - [x] Move `PublishChangesInSite` -> `PublishChangesInSite\PublishChangesInSiteCommand`
  - [x] Create `PublishChangesInSiteCommandHandler`
  - [x] Ensure that `PublishChangesInSiteCommandHandler` throws `ConflictsOccurred`
- [x] Turn `ConflictsOccurred` into a Result DTO rather than an exception
- [x] Eliminate redundancy in `catch (WorkspaceRebaseFailed $e)` clauses
- [x] Trigger conflict resolution from within `Publish` saga 
  - [x] Add `conflicts` case to `PublishingResponse`
  - [x] Add `PublishingState.CONFLICTS`
- [x] Fix: "Node could not be published, because of missing parentNode" after conflict resolution (happens when you do "Publish Document" on a document that gets removed by conflict resolution)
- [x] Fix: TypeError node.children is undefined (regression after https://github.com/neos/neos-ui/pull/3756 - Race Condition)
- [x] Fix: You still can get stuck in an undefined state in which the saga ceases to continue (happens when the in-between sync is cancelled)
- [x] Fix: #3785

grebaldi committed:
TASK: Move `Conflicts` concept to `Application\Shared` namespace

This moves the classes `ConflictsOccurred`, `Conflict`, `Conflicts`,
`ConflictsBuilder`, `ReasonForConflict`, `TypeOfChange` and `IconLabel`
into the `Application\Shared` namespace and adjusts all references
accordingly.

These classes are going to be reused by multiple command handlers.

grebaldi committed:
TASK: Add command handlers for all Publish-related commands

This includes the following tasks;
- Move `PublishChangesInDocument` ->
`PublishChangesInDocument\PublishChangesInDocumentCommand`
- Create `PublishChangesInDocumentCommandHandler`
- Move `PublishChangesInSite` ->
`PublishChangesInSite\PublishChangesInSiteCommand`
- Create `PublishChangesInSiteCommandHandler`

grebaldi committed:
TASK: Ensure that all Publish-related command handlers throw `ConflictsOccurred`

grebaldi committed:
TASK: Turn `ConflictsOccurred` into a Result DTO rather than an exception

grebaldi committed:
TASK: Move logic to create Conflicts from WorkspaceRebaseFailed into dedicated Factory

grebaldi committed:
TASK: Use conflict DTO `key` property as key for iteration in ConflictList component

grebaldi committed:
TASK: Trigger conflict resolution from within `Publish` saga

mhsdesign commented:
Id like to help reviewing this so we can soon make the changes Neos requires regarding the rename an removal of `Node` fields on 9.0-dev (whithout major conflicts against your pr)

grebaldi commented:
@mhsdesign Don't worry about conflicts - I'll take care of keeping this PR up-to-date. So, unless anything upstream of the UI needs this change to move on, I'd say that everything else has more priority.

grebaldi committed:
BUGFIX: Remove race condition in `watchNodeInformationChanges`

It may happen that `fetchAdditionalNodeMetadata` adds metadata for
a node that has since been removed from the store. The occurance
of such "zombie" nodes leads to problems, because they may lack
crucial properties like `children` that are treated as always-
present by other mechanisms throughout the UI.

This became apparent after https://github.com/neos/neos-ui/pull/3756.

The CR.Nodes.MERGE action now takes care of preventing zombie nodes
from entering the store.

grebaldi committed:
BUGFIX: Wait and reload before syncing in E2E test

This is to prevent the flakiness described in #3785.

grebaldi committed:
TASK: Stabilize E2E tests for synving feature

fixes: #3785

grebaldi committed:
BUGFIX: Fix undefined state when cancelling sync operations

This adds an e2e test case for cancelling and reselecting a resolution strategy
and another e2e test case for immediate resolution cancellation.

Both test cases are fixed by changes to the syncing saga.

grebaldi committed:
TASK: Handle edge case when automatic syncing during "publish document" removes the document

This also adds two more E2E test cases for publishing with automatic syncing.

grebaldi commented:
@pKallert 
FYI: While rebasing, I encountered conflicts with the fix you've provided in https://github.com/neos/neos-ui/pull/3634. To resolve this, I have moved the translation handling from `BackendServiceController::publishChangesInDocumentAction` to `PublishChangesInDocumentCommandHandler::handle`:
https://github.com/neos/neos-ui/blob/faee8cad95b72ad96e2541a0b40e09d28645393b/Classes/Application/PublishChangesInDocument/PublishChangesInDocumentCommandHandler.php#L62-L74 

Does that match your intend?

@mhsdesign 
One of the conflicting commits was one of yours, but was introduced in #3634 as well: https://github.com/neos/neos-ui/commit/29a73a166afe1f3e2a545a577494b62875262785

So, same question goes to you as well :)

mhsdesign commented:
Yes looks perfect thanks for taking care :)

mhsdesign commented:
Hi i kindof forget about this ^^ Would this be something for the next beta to include?

And while navigating the code i found that this file is no longer in use `packages/neos-ui-redux-store/src/UI/SyncWorkspaceModal/index.ts` so we might as well delete that? Seems its a leftover from paulas previous pr?

mhsdesign commented on Classes/Infrastructure/ContentRepository/ConflictsFactory.php:
```diff
@@ -127,6 +114,9 @@ private function createConflictFromCommandThatFailedDuringRebase(
         return new Conflict(
+            key: $affectedNode
+                ? $affectedNode->nodeAggregateId->value
+                : 'command-' . $commandThatFailedDuringRebase->sequenceNumber,
```
this is some kind of deduplication thing i think but i dont fully understand it yet :D

mhsdesign commented on Classes/Infrastructure/ContentRepository/ConflictsFactory.php:
```diff
@@ -32,72 +32,59 @@
+    public function fromWorkspaceRebaseFailed(
+        WorkspaceRebaseFailed $workspaceRebaseFailed
+    ): Conflicts {
+        $conflictsBuilder = Conflicts::builder();
```
i know you like builders, but id be fine by using and array to track it ^^

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/conflict-resolution-03/rebase-during-publish

mhsdesign committed:
TASK: Fix php code

mhsdesign commented on Classes/Infrastructure/ContentRepository/ConflictsFactory.php:
```diff
@@ -127,6 +114,9 @@ private function createConflictFromCommandThatFailedDuringRebase(
 
         return new Conflict(
+            key: $affectedNode
+                ? $affectedNode->nodeAggregateId->value
```
```suggestion
                ? $affectedNode->aggregateId->value
```

mhsdesign reviewed (commented):
There where some changes in the e2e setup and the test dont pass all on 9.0 anymore but that might be fixed here as well. But i still happen to get publish errors when running the tests locally: And the tests fail at a later point because we just click the `Discard` button with the id `#neos-DiscardDialog-Acknowledg` even if its an error:

<img width="933" alt="image" src="https://github.com/user-attachments/assets/ade785db-01d1-4505-a410-3409138231f1">

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/conflict-resolution-03/rebase-during-publish

mhsdesign committed:
TASK: Adjust to changes in 9.0 (workspace metadata)

mhsdesign committed:
TASK: Dare to unskip tests again:)

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/conflict-resolution-03/rebase-during-publish

mhsdesign commented:
That rebasing doesnt work is actually a core bug and will be resolved via: https://github.com/neos/neos-development-collection/pull/5301 ... that fixes most of the tests locally but one is still failing ...

mhsdesign committed:
TASK: Make sure e2e test fail if the Publish Result Dialog shows an error

Currently, the `#neos-PublishDialog-Acknowledge` id will be used both for the "ok" button in success, and for the "cancel" button in case of any unexpected errors.

To make the e2e tests fail fast we change the id in the error case to `Acknowledge-Error` so we dont accidentally click it.

see also: https://github.com/neos/neos-ui/pull/3769#pullrequestreview-2332466270

mhsdesign committed:
TASK: Don't absorb unexpected exception if publishing failed

solves https://github.com/neos/neos-ui/issues/3845

mhsdesign committed:
TASK: Adjust to changes of Publishing V3

https://github.com/neos/neos-development-collection/pull/5301

mhsdesign committed:
TASK: Avoid use of internal `sequenceNumber` for building commands that failed

instead we can just generate a uuid as the `sequenceNumber` was also unique for each command that failed

see https://github.com/neos/neos-development-collection/pull/5301#discussion_r1818111183

mhsdesign committed:
TASK: Inline `ConflictsBuilder` to better document deduplication logic

mhsdesign reviewed (approved):
Thank you so much ❤️ tested in combination with the new publishing v3 everything works again locally: https://github.com/neos/neos-development-collection/pull/5301

(I solved conflicts with 9.0 and also pushed a few minor changes)

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/conflict-resolution-03/rebase-during-publish
